pkg/api/v1/types.go

```diff
@@ -924,13 +940,31 @@ type VirtualMachineList struct {
 	Items []VirtualMachine `json:"items"`
 }
 
+// VirtualMachineRunStrategy is a label for the requested VirtualMachineInstance Running State at the current time.
+// ---
+// +k8s:openapi-gen=true
+type VirtualMachineRunStrategy string
+
+// These are the valid VMI running states
+const (
```

Comment:
I was under the impression we converged on not having the "runAlways" and "halted" run strategies. 'Running' would be used to indicate these options as they already are today, and then runStrategy would be used for this more nuanced and advanced behavior.

@fabiand @rmohr @stu-gott, have I misinterpreted the consensus that was reached here?

Also, I'm not 100% sure whether this would break the api or not, but if 'running' became a string value where 'true' and 'false' do the same things it already does today, it might technically be possible to convert running directly from a bool to an enum (true|false|manual|toCompletion) There may be side effects from this that are undesirable though. Maybe that was already discussed and I missed the details.

Reply:
> I was under the impression we converged on not having the "runAlways" and "halted" run strategies. 'Running' would be used to indicate these options as they already are today, and then runStrategy would be used for this more nuanced and advanced behavior.
> 
> @fabiand @rmohr @stu-gott, have I misinterpreted the consensus that was reached here?

I thought we did come to a consensus that we were going to keep all 4. For me personally, I find it much simpler to explain and understand to say "use either running or runStrategy, but not both" than it is to say "well, you should use running to do this, or runStrategy to do that". Also, please note the implementation gets a bit messy. Everywhere there's a simple switch block based on run strategies, there would either need be some spagehtti logic to sort out running vs runStrategy or there would need to be something that renders the two back into a nicely arranged set of enums, like it is now.

An even stronger implementation argument is that you can't switch from running=<true|false> to runStrategy=<restartOnError|manual> using patch (because none of JSONPatchType, MergePatchType, StrategicMergePatchType can delete an existing key)--which would break the subresource.go module. it would need to be re-implemented using update, which means two api calls instead of one, and introduces race conditions.

There is precedent for overlapping behaviors in the Kubernetes api. matchLabel is a complete subset of matchExpression, and either can be used in a selector.

If the only argument against having all 4 is "but I like the other way better", please strongly consider the case I've made. Removing runAlways and halted makes the code messier and measurably worse.

> Also, I'm not 100% sure whether this would break the api or not, but if 'running' became a string value where 'true' and 'false' do the same things it already does today, it might technically be possible to convert running directly from a bool to an enum (true|false|manual|toCompletion) There may be side effects from this that are undesirable though. Maybe that was already discussed and I missed the details.

changing running to a string *definitely* breaks the API. full stop.

`running: true`
becomes:
`running: "true"`